core/src/main/scala/org/bitcoins/core/protocol/ln/LnInvoiceTags.scala

```diff
@@ -1,98 +1,129 @@
+sealed abstract class LnInvoiceTag extends NetworkElement {
 
-  def description: Option[String]
+  def prefix: LnTagPrefix
 
-  def spk: Option[ScriptPubKey]
+  /** Converts the data portion of the LnInvoiceTag to a bech32 string */
+  def dataToBech32: String = {
+    Bech32.encode8bitToString(bytes)
+  }
 
-  def descriptionHash: Option[Sha256Digest]
+  override def toString: String = {
+    val b = new StringBuilder
+    b.append(prefix)
+    b.append(dataToBech32)
+    b.toString()
+  }
+}
 
-  def expiryTime: Option[UInt64]
+/**
+ * All of the different invoice tags that are currently defined
+ * Refer to BOLT11 for a full list
+ * [[https://github.com/lightningnetwork/lightning-rfc/blob/master/11-payment-encoding.md#tagged-fields]]
+ */
+object LnInvoiceTag {
 
-  def cltvExpiry: Option[UInt64]
+  case class PaymentHashTag(hash: Sha256Digest) extends LnInvoiceTag {
 
-  type fallbackAddress = (Int, BitcoinAddress)
-  def fallbackAddress: Option[fallbackAddress]
+    override val prefix: LnTagPrefix = LnTagPrefix.PaymentHash
 
-  def routingInfo: Option[Vector[LnRoutingInfo]] //TODO: Not implemented
+    override val bytes = hash.bytes
+  }
 
-  def toUnsignedByte(byte: Byte): Int = byte & 0xFF
+  case class DescriptionTag(string: String) extends LnInvoiceTag {
+    override val prefix: LnTagPrefix = LnTagPrefix.Description
 
-  override def toString: String = {
-    fromHexStrToBech32(LnTagPrefix.PaymentHash, paymentHash.hex) +
-      fromStringToBech32(LnTagPrefix.Description, description.getOrElse("")) +
-      fromHexStrToBech32(LnTagPrefix.SignaturePubKey, spk.map(_.hex).getOrElse("")) +
-      fromHexStrToBech32(LnTagPrefix.DescriptionHash, descriptionHash.map(_.hex).getOrElse("")) +
-      fromUInt64toBech32(LnTagPrefix.ExpiryTime, expiryTime) +
-      fromUInt64toBech32(LnTagPrefix.CltvExpiry, cltvExpiry) +
-      fallbackAddressBech32(fallbackAddress)
-    //TODO: Add routingInfo
-  }
+    override val bytes: ByteVector = ByteVector(string.getBytes("UTF-8"))
 
-  def fallbackAddressBech32(input: Option[fallbackAddress]): String = {
-    if (input.isDefined) {
-      val uInt8Array = uInt64ToBase32(UInt64(input.get._1)) ++ fromHexToBase32(input.get._2.hash.hex)
-      val bech32String = Bech32Address.encodeToString(uInt8Array)
-      val bech32DataLength = bech32EncodeDataLength(bech32String)
-      LnTagPrefix.FallbackAddress + bech32DataLength + bech32String
-    } else {
-      ""
-    }
   }
 
-  def fromStringToBech32(prefix: LnTagPrefix, tag: String): String = {
-    if (!tag.isEmpty) {
-      val uInt8Array = tag.map(a => a.toByte).map(i => UInt8(toUnsignedByte(i).toShort))
-      val base32Array = Bech32Address.encode(uInt8Array)
-      val bech32String = Bech32Address.encodeToString(base32Array.get)
-      val bech32DataLength = bech32EncodeDataLength(bech32String)
-      prefix.value + bech32DataLength + bech32String
-    } else { "" }
+  case class ScriptPubKeyTag(spk: ScriptPubKey) extends LnInvoiceTag {
```

Comment:
This is meant to be the `n` tag which denotes the "33-byte public key of the payee node." I think you are getting it confused with the `f` tag which denotes the "Fallback on-chain address: for bitcoin, this starts with a 5-bit version and contains a witness program or P2PKH or P2SH address." ?